File: optics/sanity_check.lua
require 'pl'
S4 = require 'RCWA'

function interp1d(x, y, xval,extrapolate)
    -- A function to compute linear interpolation of 1 dimensional data with extrapolation 
    --
    -- Set default values
    extrapolate = extrapolate or false

    -- If x is outside the interpolation range we need to extrapolate
    min,max = x:minmax()
    if xval < min and extrapolate then
        print('Extrapolate left')
        m = (y[2] - y[1])/(x[2] - x[1])
        val = m*(xval-x[1])+y[1]
        --print('Left: ',y[1])
        --print('Value: ',val)
        --print('Right: ',y[2])
    elseif xval > max and extrapolate then
        print('Extrapolate right')
        m = (y[#y] - y[#y-1])/(x[#x] - x[#x-1])
        val = m*(xval-x[#x])+y[#x]
        --print('Left: ',y[#y-1])
        --print('Value: ',val)
        --print('Right: ',y[#y])
    elseif (xval < min or xval > max) and not extrapolate then
        error("The x value is outside the extrapolation range and extrapolation is set to false")
    else 

        -- First find the first x value that exceeds our desired value
        counter = 1
        while x[counter] < xval do
            counter = counter + 1
        end
        val = (((y[counter]-y[counter-1])*(xval - x[counter-1]))/(x[counter]-x[counter-1]))+y[counter-1]
        --print('Left: ',y[counter-1])
        --print('Value: ',val)
        --print('Right: ',y[counter])
    end
    return val
end

function parse_nk(path)
    -- Set path to current input file
    io.input(path)
    -- Skip the header line
    io.read()
    freqvec = List()
    nvec = List()
    kvec = List()
    -- Get the columns
    for freq,n,k in input.fields(3) do
        freqvec:append(freq)
        nvec:append(n)
        kvec:append(k)
    end
    return freqvec:reverse(),nvec:reverse(),kvec:reverse()
end

function interpolate(conf)
    -- Perform interpolation of conf objects
    
    -- For each section of parameters in the configuration table
    for sect,params in pairs(conf) do
        -- For each parameter, value pair
        for param, value in pairs(params) do
            -- Search the value of the param for the interpolation pattern
            --
            -- Matches $(some_params)s where any character EXCEPT a closing parenthesis
            -- ) can be inside the %( )s. Extracts whats inside interp string
            matches = string.gmatch(value,"%%%(([^)]+)%)s") 
            -- For each parameter name extracted from the interpolation string 
            for match in matches do
                -- Get the value of tha param name. Note this only works if the sought
                -- param is in the same section as the discovered interp string
                -- TODO: Fix this so it searches the conf table for the proper KEY recursively
                repl_val = conf[sect][match] 
                -- Now replace the interpolation string by the actual value of the 
                -- sought parameter 
                rep_par = string.gsub(value,"%%%(([^)]+)%)s",repl_val)
                -- Now store the interpolated value of the original parameter
                conf[sect][param] = rep_par
            end
        end
    end
    return conf
end

function get_epsilon(freq,path)
    freqvec,nvec,kvec= parse_nk(path)
    n = interp1d(freqvec,nvec,freq,true)
    k = interp1d(freqvec,kvec,freq,true)
    eps_r = n*n - k*k
    eps_i = 2*n*k
    return eps_r,eps_i
end

function run_air_sim(p,m)
    sim = S4.NewSimulation()
    sim:SetLattice({p['L'],0},{0,p['L']})
    sim:SetNumG(p['numbasis'])
    sim:SetVerbosity(2)
    sim:SetMaterial('vacuum',{1.0,0.0})
    sim:SetLayer('air1',p['layer_t'],'vacuum')
    sim:SetLayer('air2',p['layer_t'],'vacuum')
    -- Set frequency
    --
    c = 299792458
    f_phys = p['freq'] 
    c_conv = c*1E6
    f_conv = f_phys/c_conv
    print('f_phys = ',f_phys)
    print('wavelength = ',(c/f_phys)*1E6)
    print('f_conv = ',f_conv)
    sim:SetFrequency(f_conv)
    E_mag = 1.0 
    sim:SetExcitationPlanewave({0,0},{-E_mag,0},{-E_mag,90})
    x_samp = p['x_samp'] 
    y_samp = p['y_samp'] 
    z_samp = p['z_samp']
    height = p['layer_t']*3 
    dz = height/z_samp
    for i,z in ipairs(List.range(0,height,dz)) do
        sim:GetFieldPlane(z,{x_samp,y_samp},'FileAppend','test_fields')
    end
end

function run_airito_sim(p,m)
    sim = S4.NewSimulation()
    sim:SetLattice({p['L'],0},{0,p['L']})
    sim:SetNumG(p['numbasis'])
    sim:SetVerbosity(2)
    sim:SetMaterial('vacuum',{1.0,0.0})
    sim:SetLayer('air1',p['layer_t'],'vacuum')
    eps_r,eps_i = get_epsilon(p['freq'],m['ito'])
    sim.SetMaterial('ito',{eps_r,eps_i})
    sim:SetLayer('ito',p['layer_t'],'ito')
    sim:SetLayer('air2',p['layer_t'],'vacuum')
    -- Set frequency
    --
    c = 299792458
    f_phys = p['freq'] 
    c_conv = c*1E6
    f_conv = f_phys/c_conv
    print('f_phys = ',f_phys)
    print('wavelength = ',(c/f_phys)*1E6)
    print('f_conv = ',f_conv)
    sim:SetFrequency(f_conv)
    E_mag = 1.0 
    sim:SetExcitationPlanewave({0,0},{-E_mag,0},{-E_mag,90})
    x_samp = p['x_samp'] 
    y_samp = p['y_samp'] 
    z_samp = p['z_samp']
    height = p['layer_t']*2   
    for z in List.range(0,height,height/z_samp) do
        sim:GetFieldPlane(z,{x_samp,y_samp},'FileAppend','test_fields')
    end
end

function main()
    args = lapp [[
Sanity check for S4 library
    --air  Run air simulation
    --ito  Run air-ito simulation
    --sub  Run air-ito-sub simulation
]]

    params = {freq=5E14,layer_t=.5,L=.25,x_samp=50,y_samp=50,z_samp=600,numbasis=40}
    plane = params['x_samp']/2
    params['plane'] = plane
    mats = {ito='/home/kyle_robertson/schoolwork/gradschool/nanowire/code/NK/008_ITO_nk_Hz.txt',
            gaas='/home/kyle_robertson/schoolwork/gradschool/nanowire/code/NK/006_GaAs_nk_Walker_modified_Hz.txt'}
    dir.makepath('sanity_check_run')
    basedir = path.join(path.currentdir(),'sanity_check_run')
    if args['air'] then
        pth = path.join(basedir,'air_sim')
        dir.makepath(pth)
        path.chdir(pth)
        run_air_sim(params,mats)
        print('Finished air sim')
        path.chdir(basedir)
    end
    if args['ito'] then 
        pth = path.join(basedir,'ito_sim')
        dir.makepath(pth)
        path.chdir(pth)
        run_airito_sim(params,mats)
        print('Finished ito sim')
        path.chdir(basedir)
    end
    if args['sub'] then 
        pth = path.join(basedir,'sub_sim')
        dir.makepath(pth)
        path.chdir(pth)
        run_airitosub_sim(params,mats)
        print('Finished ito sim')
        path.chdir(basedir)
    end
end
main()
-----------------------------------------------------------------------
-- NEED TO PORT ALL CODE BELO HERE
-----------------------------------------------------------------------

    

--function run_airitosub_sim(p,m)
--    sim = S4.New(Lattice=((p['L'],0),(0,p['L'])),NumBasis=p['numbasis'])
--    sim.SetVerbosity(2)
--    sim.SetMaterial(Name='vacuum',Epsilon=complex(1.0,0.0))
--    eps = get_epsilon(p['freq'],m['ito'])
--    sim.SetMaterial(Name='ito',Epsilon=eps)
--    eps = get_epsilon(p['freq'],m['gaas'])
--    sim.SetMaterial(Name='gaas',Epsilon=eps)
--    sim.AddLayer(Name='air1',Thickness=p['layer_t'],Material='vacuum')
--    sim.AddLayer(Name='ito',Thickness=p['layer_t'],Material='ito')
--    sim.AddLayer(Name='gaas',Thickness=p['layer_t'],Material='gaas')
--    sim.AddLayerCopy('air2',Thickness=p['layer_t'],Layer='air1')
--    # Set frequency
--    f_phys = p['freq'] 
--    c_conv = constants.c*1E6
--    f_conv = f_phys/c_conv
--    print('f_phys = ',f_phys)
--    print('wavelength = ',(constants.c/f_phys)*1E6)
--    print('f_conv = ',f_conv)
--    sim.SetFrequency(f_conv)
--    E_mag = 1.0 
--    sim.SetExcitationPlanewave(IncidenceAngles=(0,0),sAmplitude=complex(E_mag,0),
--            pAmplitude=complex(0,0))
--    x_samp = p['x_samp'] 
--    y_samp = p['y_samp'] 
--    z_samp = p['z_samp']
--    height = p['layer_t']*4 
--    for z in np.linspace(0,height,z_samp):
--        sim.GetFieldsOnGrid(z,NumSamples=(x_samp,y_samp),
--                            Format='FileAppend',BaseFilename='test_fields')
--end
--
--def run_airitowiresub_sim(params):
--    sim = S4.New(Lattice=((.63,0),(0,.63)),NumBasis=25)
--    sim.SetVerbosity(2)
--    sim.SetMaterial(Name='vacuum',Epsilon=complex(1.0,0.0))
--    sim.SetMaterial(Name='ito',Epsilon=complex(2.0766028416,0.100037324))
--    sim.SetMaterial(Name='gaas',Epsilon=complex(3.5384,0.0))
--    sim.SetMaterial(Name='cyc',Epsilon=complex(1.53531,1.44205E-6))
--    sim.AddLayer(Name='air1',Thickness=.5,Material='vacuum')
--    sim.AddLayer(Name='ito',Thickness=.5,Material='ito')
--    sim.AddLayer(Name='wire',Thickness=.5,Material='cyc')
--    sim.SetRegionCircle('wire','gaas',(0,0),.2)
--    sim.AddLayer(Name='gaas',Thickness=.5,Material='gaas')
--    # Set frequency
--    f_phys = 3E14 
--    c_conv = constants.c/10E-6
--    f_conv = f_phys/c_conv
--    print('f_phys = ',f_phys)
--    print('wavelength = ',(constants.c/f_phys)*10E6)
--    print('f_conv = ',f_conv)
--    sim.SetFrequency(f_conv)
--    E_mag = 1.0 
--    sim.SetExcitationPlanewave(IncidenceAngles=(0,0),sAmplitude=complex(E_mag,0), pAmplitude=complex(0,E_mag))
--    x_samp = 200 
--    y_samp = 200 
--    z_samp = 200
--    height = 2.0 
--    for z in np.linspace(0,height,z_samp):
--        sim.GetFieldsOnGrid(z,NumSamples=(x_samp,y_samp),
--                            Format='FileAppend',BaseFilename='test_fields')

File: optics/utils/sanity_check.py
#import S4
import numpy as np
from scipy import interpolate
from scipy import constants
import matplotlib
import matplotlib.pyplot as plt
from mpl_toolkits.mplot3d import Axes3D
import matplotlib.cm as cmx
import argparse as ap
import os

def get_epsilon(freq,path):
    """Returns complex dielectric constant for a material by pulling in nk text file, interpolating,
    computing nk values at freq, and converting"""
    # Get data
    freq_vec,n_vec,k_vec = np.loadtxt(path,skiprows=1,unpack=True)
    # Get n and k at specified frequency via interpolation 
    f_n = interpolate.interp1d(freq_vec,n_vec)
    f_k = interpolate.interp1d(freq_vec,k_vec)
    n,k = f_n(freq),f_k(freq)
    # Convert to dielectric constant
    # NOTE!!: This assumes the relative magnetic permability (mew) is 1
    epsilon_real = n**2 - k**2
    epsilon_imag = 2*n*k
    epsilon = complex(epsilon_real,epsilon_imag)
    # NOTE: Include below in some sort of unit test later
    #print("n = %f"%n)
    #print("k = %f"%k)
    #print('epsilon = {:.2f}'.format(epsilon))
    #plt.plot(freq_vec,k_vec,'bs',freq_vec,f_k(freq_vec),'r--')
    #plt.show()
    return epsilon

def calcnormE(data):
    #Ex = data[:,3] + 1j*data[:,6]
    #Ey = data[:,4] + 1j*data[:,7]
    #Ez = data[:,5] + 1j*data[:,8]
    #E_mag = np.sqrt(Ex*np.conj(Ex)+Ey*np.conj(Ey)+Ez*np.conj(Ez))
    E_mag = np.zeros_like(data[:,3])
    for i in range(3,9):
        E_mag += data[:,i]*data[:,i]
    E_mag = np.sqrt(E_mag)
    new_data = np.zeros((data.shape[0],data.shape[1]+1))
    new_data[:,:-1] = data
    new_data[:,-1] = E_mag
    return new_data, E_mag

def comp_conv(val):
    if str(val).strip("b'") == 'NaN':
        return float(0)
    else:
        try:
            return float(val)
        except ValueError:
            val = str(val).strip("b'")
            val = val.replace('i','j')
            return complex(val)

def get_comsol(path):
    conv = {col: comp_conv for col in range(0,2)}
    raw = np.loadtxt(path,comments='#',converters=conv,dtype=complex)
    return raw

def plot_sim(p,save):
    # Load the data
    coms_dat = get_comsol('/home/kyle_robertson/schoolwork/gradschool/nanowire/code/optics/comsol_data_fixed.txt')
    data = np.loadtxt('test_fields.E')
    mat, normE = calcnormE(data)
    np.savetxt('processed_data.txt',data)
    ## Planar plot of norm of E
    pval = p['plane']
    #mat = np.column_stack((data,normE))
    #planes = np.array([row for row in mat if row[0] == pval])
    #dx = p['L']/p['x_samp']
    #dy = p['L']/p['y_samp']
    #x,y,z = np.unique(planes[:,0])*dx,np.unique(planes[:,1])*dy,np.unique(planes[:,2])
    #normE = planes[:,-1].reshape(z.shape[0],y.shape[0])
    #colorsMap = 'jet' 
    #cm = plt.get_cmap(colorsMap)
    #cNorm = matplotlib.colors.Normalize(vmin=np.amin(normE), vmax=np.amax(normE))
    #scalarMap = cmx.ScalarMappable(norm=cNorm, cmap=cm)
    #fig = plt.figure(1,figsize=(9,7))
    #ax = fig.add_subplot(111)
    #ax.pcolormesh(y, z, normE,cmap=cm,norm=cNorm,alpha=.5)
    #scalarMap.set_array(normE)
    #cb = fig.colorbar(scalarMap)
    #labels = ('y','z','normE')
    #cb.set_label(labels[-1])
    #ax.set_xlabel(labels[0])
    #ax.set_ylabel(labels[1])
    #start, end = ax.get_xlim()
    #ax.xaxis.set_ticks(np.arange(start,end,0.1))
    #start, end = ax.get_ylim()
    #ax.yaxis.set_ticks(np.arange(start,end,0.1))
    #fig.suptitle('Electric field Norm') 
    #if save:
    #    plt.savefig('normE_plane_plot.pdf')
    #plt.show()
    # Plot along line 
    line = np.array([row for row in mat if row[0] == pval and row[1] == pval])
    #plt.figure(2)
    #plt.plot(line[:,2],line[:,-1],'b-')
    #plt.xlabel('z [um]')
    #plt.ylabel('norm E')
    #plt.title('Norm of Electric Field along Line')
    #plt.show()
    #if save:
    #    plt.savefig('normE_line_plot.pdf')
    # Plot real and imaginary components
    fig, axes = plt.subplots(2,2)
    axes[0,0].plot(line[:,2],line[:,3],'r-',label='Ex Real S4')
    axes[0,0].plot(line[:,2],line[:,5],'b-',label='Ey Real S4')
    axes[0,0].legend()
    axes[0,1].plot(line[:,2],line[:,4],'g-',label='Ex Imag S4')
    axes[0,1].plot(line[:,2],line[:,6],'m-',label='Ey Imag S4')
    axes[0,1].legend()
    axes[1,0].plot(range(coms_dat.shape[0]),np.real(coms_dat[:,0]),'r-',label='Ex Real COMS')
    axes[1,0].plot(range(coms_dat.shape[0]),np.real(coms_dat[:,1]),'b-',label='Ey Real COMS')
    axes[1,0].legend()
    axes[1,1].plot(range(coms_dat.shape[0]),np.imag(coms_dat[:,0]),'g-',label='Ex Imag COMS')
    axes[1,1].plot(range(coms_dat.shape[0]),np.imag(coms_dat[:,1]),'m-',label='Ey Imag COMS')
    axes[1,1].legend()
    fig.suptitle('Real and Imag Components on Line')
    if save:
        plt.savefig('real_imag_line.pdf')
    plt.show()
    # Plot components magnitudes 
    #plt.figure(3)
    #ex_mag = np.sqrt(line[:,3]*line[:,3]+line[:,4]*line[:,4])
    #ey_mag = np.sqrt(line[:,5]*line[:,5]+line[:,6]*line[:,6])
    #plt.plot(line[:,2],ex_mag,'r-',label='Ex Magnitude')
    #plt.plot(line[:,2],ey_mag,'b-',label='Ey Magnitude')
    #plt.legend()
    #plt.title('Magnitudes on Line')
    #if save:
    #    plt.savefig('real_imag_line.pdf')
    #plt.show()
def run_air_sim(p,m):
    sim = S4.New(Lattice=((p['L'],0),(0,p['L'])),NumBasis=p['numbasis'])
    sim.SetOptions(Verbosity=2)
    sim.SetMaterial(Name='vacuum',Epsilon=complex(1.0,0.0))
    sim.AddLayer(Name='air1',Thickness=p['layer_t'],Material='vacuum')
    sim.AddLayer(Name='air2',Thickness=p['layer_t'],Material='vacuum')
    # Set frequency
    f_phys = p['freq'] 
    c_conv = constants.c*1E6
    f_conv = f_phys/c_conv
    print('f_phys = ',f_phys)
    print('wavelength = ',(constants.c/f_phys)*1E6)
    print('f_conv = ',f_conv)
    sim.SetFrequency(f_conv)
    E_mag = 1.0 
    sim.SetExcitationPlanewave(IncidenceAngles=(0,0),sAmplitude=complex(E_mag,0),
            pAmplitude=complex(0,-E_mag))
    x_samp = p['x_samp'] 
    y_samp = p['y_samp'] 
    z_samp = p['z_samp']
    height = p['layer_t']*2   
    for z in np.linspace(0,height,z_samp):
        sim.GetFieldsOnGrid(z,NumSamples=(x_samp,y_samp),
                            Format='FileAppend',BaseFilename='test_fields')

def run_airito_sim(p,m):
    sim = S4.New(Lattice=((p['L'],0),(0,p['L'])),NumBasis=p['numbasis'])
    sim.SetOptions(Verbosity=2)
    sim.SetMaterial(Name='vacuum',Epsilon=complex(1.0,0.0))
    eps = get_epsilon(p['freq'],m['ito'])
    sim.SetMaterial(Name='ito',Epsilon=eps)
    sim.AddLayer(Name='air1',Thickness=p['layer_t'],Material='vacuum')
    sim.AddLayer(Name='ito',Thickness=p['layer_t'],Material='ito')
    sim.AddLayerCopy('air2',Thickness=p['layer_t'],Layer='air1')
    # Set frequency
    f_phys = p['freq'] 
    c_conv = constants.c*1E6
    f_conv = f_phys/c_conv
    print('f_phys = ',f_phys)
    print('wavelength = ',(constants.c/f_phys)*1E6)
    print('f_conv = ',f_conv)
    sim.SetFrequency(f_conv)
    E_mag = 1.0 
    sim.SetExcitationPlanewave(IncidenceAngles=(0,0),sAmplitude=complex(E_mag,0),
            pAmplitude=complex(0,E_mag))
    x_samp = p['x_samp'] 
    y_samp = p['y_samp'] 
    z_samp = p['z_samp']
    height = p['layer_t']*3  
    for z in np.linspace(0,height,z_samp):
        sim.GetFieldsOnGrid(z,NumSamples=(x_samp,y_samp),
                            Format='FileAppend',BaseFilename='test_fields')

def run_airitosub_sim(p,m):
    sim = S4.New(Lattice=((p['L'],0),(0,p['L'])),NumBasis=p['numbasis'])
    sim.SetOptions(Verbosity=2)
    sim.SetMaterial(Name='vacuum',Epsilon=complex(1.0,0.0))
    eps = get_epsilon(p['freq'],m['ito'])
    sim.SetMaterial(Name='ito',Epsilon=eps)
    eps = get_epsilon(p['freq'],m['gaas'])
    sim.SetMaterial(Name='gaas',Epsilon=eps)
    sim.AddLayer(Name='air1',Thickness=p['layer_t'],Material='vacuum')
    sim.AddLayer(Name='ito',Thickness=p['layer_t'],Material='ito')
    sim.AddLayer(Name='gaas',Thickness=p['layer_t'],Material='gaas')
    sim.AddLayerCopy('air2',Thickness=p['layer_t'],Layer='air1')
    # Set frequency
    f_phys = p['freq'] 
    c_conv = constants.c*1E6
    f_conv = f_phys/c_conv
    print('f_phys = ',f_phys)
    print('wavelength = ',(constants.c/f_phys)*1E6)
    print('f_conv = ',f_conv)
    sim.SetFrequency(f_conv)
    E_mag = 1.0 
    sim.SetExcitationPlanewave(IncidenceAngles=(0,0),sAmplitude=complex(E_mag,0),
            pAmplitude=complex(0,0))
    x_samp = p['x_samp'] 
    y_samp = p['y_samp'] 
    z_samp = p['z_samp']
    height = p['layer_t']*4 
    for z in np.linspace(0,height,z_samp):
        sim.GetFieldsOnGrid(z,NumSamples=(x_samp,y_samp),
                            Format='FileAppend',BaseFilename='test_fields')

def run_airitowiresub_sim(params):
    sim = S4.New(Lattice=((.63,0),(0,.63)),NumBasis=25)
    sim.SetOptions(Verbosity=2)
    sim.SetMaterial(Name='vacuum',Epsilon=complex(1.0,0.0))
    sim.SetMaterial(Name='ito',Epsilon=complex(2.0766028416,0.100037324))
    sim.SetMaterial(Name='gaas',Epsilon=complex(3.5384,0.0))
    sim.SetMaterial(Name='cyc',Epsilon=complex(1.53531,1.44205E-6))
    sim.AddLayer(Name='air1',Thickness=.5,Material='vacuum')
    sim.AddLayer(Name='ito',Thickness=.5,Material='ito')
    sim.AddLayer(Name='wire',Thickness=.5,Material='cyc')
    sim.SetRegionCircle('wire','gaas',(0,0),.2)
    sim.AddLayer(Name='gaas',Thickness=.5,Material='gaas')
    # Set frequency
    f_phys = 3E14 
    c_conv = constants.c/10E6
    f_conv = f_phys/c_conv
    print('f_phys = ',f_phys)
    print('wavelength = ',(constants.c/f_phys)*10E6)
    print('f_conv = ',f_conv)
    sim.SetFrequency(f_conv)
    E_mag = 1.0 
    sim.SetExcitationPlanewave(IncidenceAngles=(0,0),sAmplitude=complex(E_mag,0), pAmplitude=complex(0,E_mag))
    x_samp = 200 
    y_samp = 200 
    z_samp = 200
    height = 2.0 
    for z in np.linspace(0,height,z_samp):
        sim.GetFieldsOnGrid(z,NumSamples=(x_samp,y_samp),
                            Format='FileAppend',BaseFilename='test_fields')

def main():
    parser = ap.ArgumentParser(description="""Sanity check for S4 library""")
    parser.add_argument('--air',action='store_true',help="Run air simulation")
    parser.add_argument('--ito',action='store_true',help="Run air-ito simulation")
    parser.add_argument('--sub',action='store_true',help="Run air-ito-substrate simulation")
    parser.add_argument('--save_plots',action='store_true',help="Save all generated plots")
    parser.add_argument('--plot',type=str,default='air',help="Plot stuff")
    args = parser.parse_args()

    params = {'freq':5E14,'layer_t':.5,'L':.25,'x_samp':50,'y_samp':50,'z_samp':600,'numbasis':40}
    plane = params['x_samp']/2
    params['plane'] = plane
    mats = {'ito':'/home/kyle_robertson/schoolwork/gradschool/nanowire/code/NK/008_ITO_nk_Hz.txt',
            'gaas':'/home/kyle_robertson/schoolwork/gradschool/nanowire/code/NK/006_GaAs_nk_Walker_modified_Hz.txt'}
    #os.mkdir('sanity_check_run')
    basedir = os.path.join(os.getcwd(),'sanity_check_run')
    if args.air:
        path = os.path.join(basedir,'air_sim')
        os.mkdir(path)
        os.chdir(path)
        run_air_sim(params,mats)
        print('Finished air sim')
        plot_sim(params,args.save_plots)
        os.chdir(basedir)
    if args.ito:
        path = os.path.join(basedir,'ito_sim')
        os.mkdir(path)
        os.chdir(path)
        run_airito_sim(params,mats)
        print('Finished ito sim')
        plot_sim(params,args.save_plots)
    if args.sub:
        path = os.path.join(basedir,'sub_sim')
        os.mkdir(path)
        os.chdir(path)
        run_airitosub_sim(params,mats)
        print('Finished ito sim')
        plot_sim(params,args.save_plots)
    if args.plot:
        print('Plotting {}'.format(args.plot))
        path = os.path.join(basedir,args.plot+'_sim')
        os.chdir(path)
        plot_sim(params,args.save_plots)

if __name__ == '__main__':
    main()
